### PR TITLE
FEATURE: CL-8490 gooddata-js goes Fetch API

### PR DESCRIPTION
**Pull request by voy**
Todos:
- [x] add missing XHR tests
- [x] adjust other modules to new XHR api
- [x] setup running XHR tests in node
- [x] remove `bower`ed jquery
- [x] setup tests in node (without karma)
- [x] get rid of `r.ok ? r.json() : r`
- [x] ~~?rename xhr module to request?~~
- [x] update examples

### CODE REVIEW

**Comment by yenkins-admin** [comment]
Test FAILed.
Refer to this link for build results: https://ci.intgdc.com/job/gooddata-js-pull-request-builder/28/

**Comment by yenkins-admin** [comment]
Test FAILed.
Refer to this link for build results: https://ci.intgdc.com/job/gooddata-js-pull-request-builder/29/

File: webpack.test.config.js [hunk 1]
```diff
@@ -11,6 +13,9 @@ module.exports = {
 
         modulesDirectories: ['src', 'node_modules']
     },
+    plugins: [
```
Comment (akloboucnik) [inline webpack.test.config.js]: Maybe leftover from trying to make tests run, try to remove.

**Comment by yenkins-admin** [comment]
Test FAILed.
Refer to this link for build results: https://ci.intgdc.com/job/gooddata-js-pull-request-builder/30/

**Comment by yenkins-admin** [comment]
Test FAILed.
Refer to this link for build results: https://ci.intgdc.com/job/gooddata-js-pull-request-builder/31/

**Comment by yenkins-admin** [comment]
Test FAILed.
Refer to this link for build results: https://ci.intgdc.com/job/gooddata-js-pull-request-builder/32/

**Comment by yenkins-admin** [comment]
Test FAILed.
Refer to this link for build results: https://ci.intgdc.com/job/gooddata-js-pull-request-builder/33/

**Comment by yenkins-admin** [comment]
Test FAILed.
Refer to this link for build results: https://ci.intgdc.com/job/gooddata-js-pull-request-builder/34/

**Comment by yenkins-admin** [comment]
Test FAILed.
Refer to this link for build results: https://ci.intgdc.com/job/gooddata-js-pull-request-builder/35/

**Comment by yenkins-admin** [comment]
Test FAILed.
Refer to this link for build results: https://ci.intgdc.com/job/gooddata-js-pull-request-builder/36/

**Comment by yenkins-admin** [comment]
Test FAILed.
Refer to this link for build results: https://ci.intgdc.com/job/gooddata-js-pull-request-builder/37/

**Comment by yenkins-admin** [comment]
Test FAILed.
Refer to this link for build results: https://ci.intgdc.com/job/gooddata-js-pull-request-builder/38/

**Comment by yenkins-admin** [comment]
Test FAILed.
Refer to this link for build results: https://ci.intgdc.com/job/gooddata-js-pull-request-builder/39/

**Comment by yenkins-admin** [comment]
Test FAILed.
Refer to this link for build results: https://ci.intgdc.com/job/gooddata-js-pull-request-builder/40/

**Comment by yenkins-admin** [comment]
Test FAILed.
Refer to this link for build results: https://ci.intgdc.com/job/gooddata-js-pull-request-builder/41/

**Comment by yenkins-admin** [comment]
Test FAILed.
Refer to this link for build results: https://ci.intgdc.com/job/gooddata-js-pull-request-builder/42/

**Comment by yenkins-admin** [comment]
Test FAILed.
Refer to this link for build results: https://ci.intgdc.com/job/gooddata-js-pull-request-builder/43/

**Comment by yenkins-admin** [comment]
Test FAILed.
Refer to this link for build results: https://ci.intgdc.com/job/gooddata-js-pull-request-builder/44/

**Comment by yenkins-admin** [comment]
Test FAILed.
Refer to this link for build results: https://ci.intgdc.com/job/gooddata-js-pull-request-builder/45/

**Comment by yenkins-admin** [comment]
Test FAILed.
Refer to this link for build results: https://ci.intgdc.com/job/gooddata-js-pull-request-builder/46/

**Comment by yenkins-admin** [comment]
Test FAILed.
Refer to this link for build results: https://ci.intgdc.com/job/gooddata-js-pull-request-builder/47/

**Comment by yenkins-admin** [comment]
Test FAILed.
Refer to this link for build results: https://ci.intgdc.com/job/gooddata-js-pull-request-builder/48/

**Comment by yenkins-admin** [comment]
Test PASSed.
Refer to this link for build results: https://ci.intgdc.com/job/gooddata-js-pull-request-builder/49/

**Comment by gdgate** [comment]
Build succeeded.

- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul/177/ : SUCCESS in 1m 58s

**Comment by gdgate** [comment]
Build succeeded.

- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul/178/ : SUCCESS in 2m 39s

**Comment by gdgate** [comment]
Build succeeded.

- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul/180/ : SUCCESS in 2m 03s

**Comment by gdgate** [comment]
Build succeeded.

- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul/182/ : SUCCESS in 1m 58s

**Comment by gdgate** [comment]
Build succeeded.

- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul/183/ : SUCCESS in 1m 58s

**Comment by gdgate** [comment]
Build succeeded.

- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul/184/ : SUCCESS in 1m 54s

**Comment by gdgate** [comment]
Build succeeded.

- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul/185/ : SUCCESS in 1m 57s

**Comment by yenkins-admin** [comment]
Merged build triggered.

File: src/metadata.js [hunk 2]
```diff
@@ -22,19 +22,10 @@ import { getIn } from './util';
...
-    const fns = elementUris.map(function mapUrisToRequests(uri) {
-        return ajax(uri);
-    });
-
-    $.when.apply(this, fns).done(function requestsDone() {
-        // arguments is the array of resolved
-        const args = Array.prototype.slice.call(arguments);
-
-        const enriched = args.map(function mapArgumentsToObjects(element) {
+    Promise.all(fns).then((...args) => { // TODO add map to r.json()
```
Comment (vaclavbohac) [inline src/metadata.js]: Missing return?

File: src/metadata.js [hunk 3]
```diff
@@ -93,7 +80,7 @@ export function getElementDetails(elementUris) {
 * @return {Array} An array of attribute objects
 */
 export function getAttributes(projectId) {
-    return get('/gdc/md/' + projectId + '/query/attributes').then(getIn('query.entries'));
+    return get('/gdc/md/' + projectId + '/query/attributes').then(r => r.ok ? r.json() : r).then(getIn('query.entries'));
```
Comment (vaclavbohac) [inline src/metadata.js]: Seems `r => r.ok ? r.json() : r` is quite recurring here, maybe we could have something like `.then(getJSON)`

Comment (vaclavbohac) [inline src/metadata.js]: Or perhaps
```js
return getJSON(`/gdc/md/${projectId}/query/attributes`).then(getIn('query.entries'));
```

File: src/metadata.js [hunk 4]
```diff
@@ -131,15 +118,12 @@ export function getFolders(projectId, type) {
...
-            /*eslint-disable new-cap*/
-            const d = $.Deferred();
-            /*eslint-enable new-cap*/
-            $.when(_getFolders(projectId, 'fact'),
-                    _getFolders(projectId, 'metric'),
-                    getDimensions(projectId)).done(function requestsDone(facts, metrics, attributes) {
-                d.resolve({fact: facts, metric: metrics, attribute: attributes});
+            return Promise.all([_getFolders(projectId, 'fact'),
+                         _getFolders(projectId, 'metric'),
+                         getDimensions(projectId)]) // TODO add r.json step
```
Comment (vaclavbohac) [inline src/metadata.js]: Forgotten todo?

**Comment by gdgate** [comment]
Build failed.

- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul/239/ : FAILURE in 1m 41s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/4/ : FAILURE in 1m 07s (non-voting)

**Comment by gdgate** [comment]
Build failed.

- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul/240/ : FAILURE in 1m 46s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/5/ : FAILURE in 1m 04s (non-voting)

**Comment by gdgate** [comment]
Build failed.

- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul/241/ : FAILURE in 1m 56s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/6/ : FAILURE in 1m 17s (non-voting)

**Comment by gdgate** [comment]
Build succeeded.

- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul/242/ : SUCCESS in 1m 03s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/7/ : FAILURE in 1m 14s (non-voting)

**Comment by gdgate** [comment]
Build succeeded.

- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul/243/ : SUCCESS in 2m 07s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/8/ : FAILURE in 1m 23s (non-voting)

**Comment by gdgate** [comment]
Build succeeded.

- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul/244/ : SUCCESS in 2m 01s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/9/ : FAILURE in 1m 17s (non-voting)

**Comment by gdgate** [comment]
Build succeeded.

- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul/245/ : SUCCESS in 2m 19s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/10/ : FAILURE in 21s (non-voting)

**Comment by gdgate** [comment]
Build succeeded.

- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul/249/ : SUCCESS in 2m 16s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/14/ : FAILURE in 1m 22s (non-voting)

**Comment by gdgate** [comment]
Build succeeded.

- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul/250/ : SUCCESS in 2m 09s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/15/ : FAILURE in 1m 27s (non-voting)

**Comment by gdgate** [comment]
Build succeeded.

- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul/251/ : SUCCESS in 2m 00s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/16/ : FAILURE in 19s (non-voting)

File: package.json [hunk 5]
```diff
@@ -50,6 +57,8 @@
     "webpack-dev-server": "1.16.1"
   },
   "dependencies": {
+    "es6-promise": "3.0.2",
+    "isomorphic-fetch": "^2.2.0",
```
Comment (vaclavbohac) [inline package.json]: Remove asterisk

**Comment by vaclavbohac** [comment]
ok to test

**Comment by gdgate** [comment]
Build succeeded.

- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul/254/ : SUCCESS in 2m 17s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/19/ : SUCCESS in 2m 16s (non-voting)

**Comment by gdgate** [comment]
Build succeeded.

- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul/255/ : SUCCESS in 2m 00s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/20/ : SUCCESS in 1m 27s (non-voting)

**Comment by gdgate** [comment]
Build failed.

- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul/256/ : FAILURE in 2m 09s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/21/ : FAILURE in 21s (non-voting)

**Comment by gdgate** [comment]
Build succeeded.

- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul/257/ : SUCCESS in 2m 15s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/22/ : SUCCESS in 31s (non-voting)

**Comment by gdgate** [comment]
Build succeeded.

- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul/258/ : SUCCESS in 36s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/23/ : SUCCESS in 1m 31s (non-voting)

**Comment by gdgate** [comment]
Build succeeded.

- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul/259/ : SUCCESS in 38s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/24/ : SUCCESS in 24s (non-voting)

**Comment by gdgate** [comment]
Build succeeded.

- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul/261/ : SUCCESS in 40s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/26/ : SUCCESS in 39s (non-voting)

**Comment by gdgate** [comment]
Build succeeded.

- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul/262/ : SUCCESS in 38s
- https://ci.intgdc.com/job/gooddata-js-unit-tests-zuul-el7/27/ : SUCCESS in 29s (non-voting)

**Comment by akloboucnik** [comment]
🍾

**Pull request merged**